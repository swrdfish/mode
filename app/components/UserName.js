import React from 'react'


class UserName extends React.Component {
	constructor(props) {
		super(props)
		this.state = {editMode: 'OFF',
					  value: 'connecting..'}

		//bind event handlers to this
		this.handleSpanOnClick = this.handleSpanOnClick.bind(this)
		this.handleChange = this.handleChange.bind(this)
		this.handleOnBlur = this.handleOnBlur.bind(this)
		this.handleKeyPress = this.handleKeyPress.bind(this)
	}

	static getDerivedStateFromProps(nextProps) {
		return {
			value: nextProps.value?nextProps.value:'connecting..'
		}
	}

	handleSpanOnClick() {
		this.setState({editMode: 'ON'})
	}

	handleChange(event) {
		this.setState({value: event.target.value});
	}

	handleOnBlur() {
		this.setState({editMode: 'OFF'})
	}

	handleKeyPress(event) {
		if(event.key == 'Enter'){
			this.setState({editMode: 'OFF'})
			this.props.onUpdateUsername(this.state.value)
		}
	}

	moveCaretAtEnd(event) {
		var temp_value = event.target.value
		event.target.value = ''
		event.target.value = temp_value
	}

	render() {
		if(this.state.editMode == 'OFF'){
			return (
				<span className="user-name" onClick={this.handleSpanOnClick}>
				{this.state.value} 
				&nbsp;&nbsp;&nbsp;&nbsp;<span className="user-name-edit fa  fa-pencil" />
				</span>
			)	
		}
		return (
			<input 
				className="user-name"
				autoFocus
				value={this.state.value}
				onChange={this.handleChange}
				onBlur={this.handleOnBlur}
				onKeyPress={this.handleKeyPress}
				onFocus={this.moveCaretAtEnd}
				type="text" />
		)
	}
}

export default UserName