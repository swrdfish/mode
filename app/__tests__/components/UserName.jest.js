import React from 'react'
import { shallow } from 'enzyme'
import UserName from '../../components/UserName'
import renderer from 'react-test-renderer';


describe('Component: UserName', () => {
    it('should render the user name', () => {
        const onUpdateUsername = jest.fn()
        const wrapper = shallow(<UserName value="foobar" onUpdateUsername={ onUpdateUsername } />);
        expect(wrapper.find(".user-name")).toHaveLength(1)
    })

    it('should show connecting in place of username if username is undefined', () => {
        const onUpdateUsername = jest.fn()
        const wrapper = shallow(<UserName onUpdateUsername={ onUpdateUsername } />);
        expect(wrapper.find(".user-name").text()).toEqual('connecting..    ')
    })

    it('should show username properly', () => {
        const onUpdateUsername = jest.fn()
        const wrapper = shallow(<UserName value="foobar" onUpdateUsername={ onUpdateUsername } />);
        expect(wrapper.find(".user-name").text()).toEqual('foobar    ') 
    })

    it('should match previous snapshot', () => {
        const onUpdateUsername = jest.fn()

        const usernameWithValidName = renderer
            .create(<UserName value="foobar" onUpdateUsername={ onUpdateUsername } />)
            .toJSON();
        expect(usernameWithUndefinedName).toMatchSnapshot();
        const usernameWithUndefinedName = renderer
            .create(<UserName onUpdateUsername={ onUpdateUsername } />)
    })
})