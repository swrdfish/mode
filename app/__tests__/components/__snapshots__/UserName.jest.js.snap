// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component: UserName should match previous snapshot 1`] = `undefined`;
